torch_np/_binary_ufuncs.py

```diff
@@ -77,69 +77,57 @@ def matmul(
 
     # NB: do not broadcast input tensors against the out=... array
     result = _binary_ufuncs.matmul(*tensors)
-    return _helpers.result_or_out(result, out)
-
-
-#
-# For each torch ufunc implementation, decorate and attach the decorated name
-# to this module. Its contents is then exported to the public namespace in __init__.py
-#
-for name in __all__:
-    ufunc = getattr(_binary_ufuncs, name)
-    decorated = normalizer(deco_binary_ufunc(ufunc))
-
-    decorated.__qualname__ = name  # XXX: is this really correct?
-    decorated.__name__ = name
-    vars()[name] = decorated
-
-
-# a stub implementation of divmod, should be improved after
-# https://github.com/pytorch/pytorch/issues/90820 is fixed in pytorch
-#
-# Implementation details: we just call two ufuncs which have been created
-# just above, for x1 // x2 and x1 % x2.
-# This means we are normalizing x1, x2 in each of the ufuncs --- note that there
-# is no @normalizer on divmod.
+    return result
 
 
 def divmod(
-    x1,
-    x2,
+    x1: ArrayLike,
+    x2: ArrayLike,
+    out1: Optional[NDArray] = None,
+    out2: Optional[NDArray] = None,
     /,
-    out=None,
+    out: Optional[tuple[NDArray]] = (None, None),
     *,
     where=True,
     casting="same_kind",
     order="K",
-    dtype=None,
+    dtype: DTypeLike = None,
     subok: SubokLike = False,
     signature=None,
     extobj=None,
 ):
-    out1, out2 = None, None
+    num_outs = sum(x is None for x in [out1, out2])
+    if sum_outs == 1:
+        raise ValueError("both out1 and out2 need to be provided")
+    if sum_outs != 0 and out != (None, None):
+        raise ValueError("Either provide out1 and out2, or out.")
     if out is not None:
         out1, out2 = out
+    if out1.shape != out2.shape or out1.dtype != out2.dtype:
+        raise ValueError("out1, out2 must be compatible")
 
-    kwds = dict(
-        where=where,
-        casting=casting,
-        order=order,
-        dtype=dtype,
-        subok=subok,
-        signature=signature,
-        extobj=extobj,
+    tensors = _helpers.ufunc_preprocess(
+        (x1, x2), out, True, casting, order, dtype, subok, signature, extobj
```

Comment:
As discussed `ufunc_preprocess` should not take an `out` parameter as setting an `out=` parameter should not affect the type of the computation. Or is this fixed in some other PR and we'll just need to update this one afterwards?

Reply:
It's still in the TODO pipeline. Will do after the current "stack" is done.